registry/storage/driver/s3-aws/s3.go

```diff
@@ -1251,21 +1261,70 @@ func (d *driver) getStorageClass() *string {
 	return aws.String(d.StorageClass)
 }
 
+type completedParts []*s3.CompletedPart
+
+func (a completedParts) Len() int           { return len(a) }
+func (a completedParts) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a completedParts) Less(i, j int) bool { return *a[i].PartNumber < *a[j].PartNumber }
+
+// buffer is a static size bytes buffer.
+type buffer struct {
+	data []byte
+}
+
+// NewBuffer returns a new bytes buffer from driver's memory pool.
+// The size of the buffer is static and set to params.ChunkSize.
+func (d *driver) NewBuffer() *buffer {
+	return d.pool.Get().(*buffer)
+}
+
+// ReadFrom reads as much data as it can fit in from r without growing its size.
+// It returns the number of bytes successfully read from r or error.
+func (b *buffer) ReadFrom(r io.Reader) (offset int64, err error) {
+	for len(b.data) < cap(b.data) && err == nil {
+		var n int
+		n, err = r.Read(b.data[len(b.data):cap(b.data)])
+		offset += int64(n)
+		b.data = b.data[:len(b.data)+n]
```

Comment:
Yeah, I'm not sure it's worth it.  Maybe? 🤷‍♂️ `len` and `cap` just return an unexported field from the underlying struct so I'd expect negligible gain here 🤔

Reply:
No; `len(b.data)` and `cap(b.data)` [are just accessing fields on the slice header](https://pkg.go.dev/reflect@go1.21.1#SliceHeader) which costs the same as any other struct field read. [Compiler explorer reveals](https://go.godbolt.org/z/nq8nrnE6E) that the compiler is able to elide bounds-checks for the expression `b.data[len(b.data):cap(b.data)]` which it cannot do when those values are copied to local variables. Copying the slice itself to a local variable sounds promising as a way to avoid dereferencing pointers inside the loop, but as compiler explorer shows, the generated AMD64 code immediately spills the slice header fields to the stack.

tl;dr introducing intermediate variables will actually deoptimize the function. Compilers are pretty smart these days.

Reply:
yes, sorry by "unexported fields" I meant "fields" and yeah, header is the right terminology. Thanks